File: e2e/helpers/factory.ts
import type { APIRequestContext } from '@playwright/test';
import { callMethod, createDoc, deleteDoc, getList } from './frappe';
import { appUrl } from './routes';
import type { WikiDocument, WikiSpace } from './wiki';

/**
 * Every space this factory creates is routed under this prefix, so the sweeper
 * in `global.teardown.ts` can delete leftovers from a killed run without any
 * chance of touching a real space.
 */
export const E2E_ROUTE_PREFIX = 'e2e';

/** A page to seed. Unknown keys are passed through to the Wiki Document. */
export interface PageSpec {
	title: string;
	content?: string;
	is_group?: boolean;
	is_published?: boolean;
	children?: PageSpec[];
	[field: string]: unknown;
}

/** A space to seed. Unknown keys are passed through to the Wiki Space. */
export interface SpaceSpec {
	route?: string;
	space_name?: string;
	is_published?: boolean;
	pages?: PageSpec[];
	[field: string]: unknown;
}

export interface SeededPage extends WikiDocument {
	children: SeededPage[];
}

export interface SeededSpace extends WikiSpace {
	/** The root group the Wiki Space created for itself, never a replacement. */
	rootGroup: string;
	/** Seeded pages, in spec order, each with its own `children`. */
	pages: SeededPage[];
	/** Look up a seeded page at any depth by title. Throws if absent. */
	page(title: string): SeededPage;
	/** `/wiki-app/spaces/<name>`, plus any deeper segments. */
	url(...segments: string[]): string;
}

let counter = 0;

/** A route no other spec (or parallel run) can collide with. */
export function uniqueRoute(slug = 'space'): string {
	counter += 1;
	return `${E2E_ROUTE_PREFIX}-${slug}-${counter}-${Date.now().toString(36)}`;
}

/** Flatten a page tree into breadth-first levels, keeping each node's parent. */
function levelsOf(
	pages: PageSpec[],
): { spec: PageSpec; parent: PageSpec | null }[][] {
	const levels: { spec: PageSpec; parent: PageSpec | null }[][] = [];
	let current = pages.map((spec) => ({
		spec,
		parent: null as PageSpec | null,
	}));
	while (current.length) {
		levels.push(current);
		current = current.flatMap(({ spec }) =>
			(spec.children ?? []).map((child) => ({ spec: child, parent: spec })),
		);
	}
	return levels;
}

/**
 * Creates wiki fixtures and remembers them so they can all be destroyed at once.
 *
 * Seeding costs `1 + depth + 1` requests — one to create the space, one bulk
 * insert per level of the page tree, and one read-back — rather than one per
 * document. Deeper levels cannot join the same bulk call: Wiki Document has no
 * `autoname`, and `set_new_name` discards any name we supply, so a child's
 * parent name is unknowable until its parent's insert returns.
 */
export class WikiFactory {
	private routes: string[] = [];

	constructor(private request: APIRequestContext) {}

	/**
	 * Seed a space and its page tree.
	 *
	 * Routes are left for the server to derive from the title and the ancestor
	 * chain, so seeded pages carry the same routes the UI would have produced.
	 */
	async space(spec: SpaceSpec = {}): Promise<SeededSpace> {
		const { pages = [], route, space_name, ...spaceFields } = spec;
		const spaceRoute = route ?? uniqueRoute();

		// Remembered before the create, so a request that times out server-side
		// still leaves us able to find and delete what it made.
		this.routes.push(spaceRoute);

		const space = await createDoc<WikiSpace & { root_group: string }>(
			this.request,
			'Wiki Space',
			{
				route: spaceRoute,
				space_name: space_name ?? spaceRoute,
				is_published: spec.is_published ?? true,
				...spaceFields,
			},
		);

		const namesBySpec = new Map<PageSpec, string>();
		for (const level of levelsOf(pages)) {
			const docs = level.map(({ spec: pageSpec, parent }) => {
				const { children, ...fields } = pageSpec;
				return {
					doctype: 'Wiki Document',
					is_published: true,
					content: `Content for ${pageSpec.title}`,
					...fields,
					parent_wiki_document: parent
						? namesBySpec.get(parent)
						: space.root_group,
				};
			});
			const names = await callMethod<string[]>(
				this.request,
				'frappe.client.insert_many',
				{ docs },
			);
			for (const [i, { spec: pageSpec }] of level.entries()) {
				namesBySpec.set(pageSpec, names[i]);
			}
		}

		const seeded = await this.readBack(namesBySpec, pages);
		const byTitle = new Map<string, SeededPage>();
		const index = (nodes: SeededPage[]) => {
			for (const node of nodes) {
				if (!byTitle.has(node.title)) byTitle.set(node.title, node);
				index(node.children);
			}
		};
		index(seeded);

		return {
			...space,
			route: spaceRoute,
			rootGroup: space.root_group,
			pages: seeded,
			page(title: string) {
				const found = byTitle.get(title);
				if (!found) {
					throw new Error(
						`No seeded page titled "${title}" in space ${spaceRoute}. ` +
							`Seeded: ${[...byTitle.keys()].join(', ') || '(none)'}`,
					);
				}
				return found;
			},
			url: (...segments: string[]) => appUrl('spaces', space.name, ...segments),
		};
	}

	/**
	 * One read to pick up what the server derived — routes, slugs, doc keys and
	 * auto-assigned sort orders — then rebuilt into the shape of the spec.
	 */
	private async readBack(
		namesBySpec: Map<PageSpec, string>,
		pages: PageSpec[],
	): Promise<SeededPage[]> {
		const names = [...namesBySpec.values()];
		if (!names.length) return [];

		const rows = await getList<WikiDocument>(this.request, 'Wiki Document', {
			fields: [
				'name',
				'title',
				'route',
				'slug',
				'doc_key',
				'is_group',
				'is_published',
				'sort_order',
				'parent_wiki_document',
			],
			filters: { name: ['in', names] },
			limit: 0,
		});
		const byName = new Map(rows.map((row) => [row.name, row]));

		const build = (specs: PageSpec[]): SeededPage[] =>
			specs.map((spec) => ({
				...(byName.get(namesBySpec.get(spec) as string) as WikiDocument),
				children: build(spec.children ?? []),
			}));
		return build(pages);
	}

	/**
	 * Destroy everything seeded through this factory.
	 *
	 * Spaces are resolved by route rather than by the create response, so a space
	 * the server made but whose response we never saw — a timed-out create, or a
	 * retry that seeded twice — is swept too. `Wiki Space.on_trash` cascades the
	 * document tree, revisions, sync logs and change requests, so a page a test
	 * created through the UI goes with it.
	 */
	async destroyAll(): Promise<void> {
		const routes = this.routes.splice(0);
		for (const route of routes) {
			await destroySpacesByRoute(this.request, route);
		}
	}
}

/** Delete every Wiki Space on `route`. Safe to call for a route that has none. */
export async function destroySpacesByRoute(
	request: APIRequestContext,
	route: string,
): Promise<void> {
	const found = await getList<{ name: string }>(request, 'Wiki Space', {
		fields: ['name'],
		filters: { route },
		limit: 0,
	}).catch(() => []);
	for (const space of found) {
		await deleteDoc(request, 'Wiki Space', space.name);
	}
}

File: wiki/tests/factory.py
"""Shared wiki fixtures for the Python test suite.

The counterpart of ``e2e/helpers/factory.ts``, with the same vocabulary: a
space is described in one call and its page tree is nested.

Four test modules each grew their own ``create_test_wiki_space`` /
``create_test_wiki_document`` before this existed, and they had drifted apart
on titles, on published defaults and on what they tracked. They now delegate
here and keep only their own bookkeeping.
"""

import frappe


def unique_route(prefix: str = "test-space") -> str:
	return f"{prefix}-{frappe.generate_hash(length=8)}"


def _insert(doc):
	"""Insert regardless of who the test is currently logged in as.

	``ignore_permissions`` is sticky -- ``insert`` stores it on the document --
	so a fixture inserted this way would carry a permission bypass into any
	later ``save()`` the test makes on the same object. Clear it, or a test
	asserting that a regular user *cannot* save gets a silent pass.
	"""
	doc.insert(ignore_permissions=True)
	doc.flags.ignore_permissions = False
	return doc


def make_space(pages: list[dict] | None = None, roles: list[tuple[str, str]] | None = None, **fields):
	"""Create a Wiki Space, and optionally a tree of pages under it.

	Prefer letting the space make its own root group: ``before_insert`` does it,
	and the group it makes is what ``on_trash`` cascades. An explicit
	``root_group`` is honoured -- several suites build one so they can parent
	documents before the space exists -- but then that document is yours to
	delete.
	"""
	route = fields.pop("route", None) or unique_route()
	space = frappe.get_doc(
		{
			"doctype": "Wiki Space",
			"route": route,
			"space_name": fields.pop("space_name", None) or route,
			"is_published": fields.pop("is_published", 1),
			**fields,
		}
	)
	for role, level in roles or []:
		space.append("roles", {"role": role, "permission_level": level})
	_insert(space)

	for spec in pages or []:
		make_document(parent=space.root_group, **spec)
	return space


def make_document(
	parent: str | None = None, title: str = "Test Page", children: list[dict] | None = None, **fields
):
	"""Create a Wiki Document under ``parent``, then its ``children``."""
	doc = frappe.get_doc(
		{
			"doctype": "Wiki Document",
			"title": title,
			"parent_wiki_document": parent,
			"is_group": 1 if (fields.pop("is_group", False) or children) else 0,
			"is_published": fields.pop("is_published", 1),
			"content": fields.pop("content", f"Content for {title}"),
			**fields,
		}
	)
	_insert(doc)

	for spec in children or []:
		make_document(parent=doc.name, **spec)
	return doc


class WikiFixtures:
	"""``make_space`` / ``make_document`` plus the bookkeeping to undo them.

	A space is deleted through ``Wiki Space.on_trash``, which cascades its
	document tree, revisions, revision items, sync logs and change requests --
	so a document inside a space needs no separate tracking.
	"""

	def __init__(self):
		self._spaces: list[str] = []
		self._documents: list[str] = []
		self._others: list[tuple[str, str]] = []

	def space(self, pages: list[dict] | None = None, **kwargs):
		return self.track_space(make_space(pages=pages, **kwargs))

	def document(self, parent: str | None = None, **kwargs):
		return self.track_document(make_document(parent=parent, **kwargs))

	def track_space(self, space):
		"""Adopt a space this factory did not create, so it is destroyed too."""
		self._spaces.append(space.name if hasattr(space, "name") else space)
		return space

	def track_document(self, doc):
		"""Adopt a document this factory did not create."""
		self._documents.append(doc.name if hasattr(doc, "name") else doc)
		return doc

	def track(self, doctype: str, doc):
		"""Adopt a row of any other doctype -- a legacy ``Wiki Page``, say."""
		self._others.append((doctype, doc.name if hasattr(doc, "name") else doc))
		return doc

	@staticmethod
	def snapshot_documents() -> set[str]:
		"""Every Wiki Document that exists right now. Pair with ``track_new``."""
		return set(frappe.get_all("Wiki Document", pluck="name", limit=0))

	def track_new(self, before: set[str]):
		"""Adopt every Wiki Document that appeared since ``before``.

		For code under test that creates documents itself -- the v3 migrations
		build them from legacy Wiki Pages, and the orphan pass parents them
		nowhere, so no space's on-trash cascade can reach them.
		"""
		for name in self.snapshot_documents() - before:
			self._documents.append(name)

	@staticmethod
	def _leaf_first(names: list[str]) -> list[str]:
		"""Order documents so no parent is deleted before its children.

		Insertion order is not enough: ``track_new`` adopts whatever the code
		under test created, in whatever order the query returned it, and the
		nested set refuses to delete a node that still has children.
		"""
		rows = (
			frappe.get_all(
				"Wiki Document",
				filters={"name": ["in", names]},
				fields=["name", "lft", "rgt"],
				limit=0,
			)
			if names
			else []
		)
		rows.sort(key=lambda r: (r.rgt or 0) - (r.lft or 0))
		ordered = [r.name for r in rows]
		# Anything the query no longer sees (already gone) keeps its place last.
		return ordered + [n for n in names if n not in set(ordered)]

	def destroy_all(self):
		"""Delete everything this factory made. Safe to call more than once."""
		previous_flag = frappe.flags.in_apply_merge_revision
		frappe.flags.in_apply_merge_revision = True
		try:
			for name in self._leaf_first(self._documents):
				if frappe.db.exists("Wiki Document", name):
					frappe.delete_doc("Wiki Document", name, force=True, ignore_permissions=True)
			self._documents.clear()

			for name in self._spaces:
				if frappe.db.exists("Wiki Space", name):
					frappe.delete_doc("Wiki Space", name, force=True, ignore_permissions=True)
			self._spaces.clear()

			for doctype, name in reversed(self._others):
				if frappe.db.exists(doctype, name):
					frappe.delete_doc(doctype, name, force=True, ignore_permissions=True)
			self._others.clear()

			# The deletion has to outlive the framework's post-test rollback.
			# A test that commits its own inserts (the reorder and rebuild APIs
			# commit on their own) leaves rows the rollback cannot reach, and an
			# uncommitted delete of those rows is itself rolled back -- which is
			# how these suites came to leave spaces behind.
			frappe.db.commit()  # nosemgrep: frappe-semgrep-rules.rules.frappe-manual-commit
		finally:
			frappe.flags.in_apply_merge_revision = previous_flag


class WikiFixtureMixin:
	"""Gives a TestCase a ``self.wiki`` factory that tears itself down.

	The factory is built on first use and registers its own ``addCleanup`` there
	and then, so it does not depend on a subclass's ``setUp`` remembering to
	call ``super()`` -- two classes in the Wiki Document suite do not. Cleanup
	runs after every ``tearDown`` in the chain, and still runs when the test
	fails.
	"""

	@property
	def wiki(self) -> WikiFixtures:
		fixtures = getattr(self, "_wiki_fixtures", None)
		if fixtures is None:
			fixtures = WikiFixtures()
			self._wiki_fixtures = fixtures
			self.addCleanup(fixtures.destroy_all)
		return fixtures
